ocaml/typing/typecore.ml

```diff
@@ -530,50 +563,56 @@ let mode_tailcall_argument mode =
 
 
 let mode_partial_application expected_mode =
-  { (mode_default (Value.regional_to_global expected_mode.mode)) with
+  let mode = alloc_as_value (value_to_alloc_r2g expected_mode.mode) in
+  { expected_mode with
+    mode;
     closure_context = Some Partial_application }
 
 
 let mode_trywith expected_mode =
   { expected_mode with position = RNontail }
 
 let mode_tuple mode tuple_modes =
+  let tuple_modes = List.map Value.disallow_left tuple_modes in
   { (mode_default mode) with
     tuple_modes }
 
-let mode_exact mode =
+let mode_exact mode exact =
   { (mode_default mode) with
-    exact = true }
+    exact = Some exact }
 
-let mode_argument ~funct ~index ~position_and_mode ~partial_app alloc_mode =
-  let vmode = Value.of_alloc alloc_mode in
-  if partial_app then mode_default vmode
+let mode_argument ~funct ~index ~position_and_mode ~partial_app marg =
+  let vmode , _ = Value.newvar_below (alloc_as_value marg) in
+  if partial_app then mode_default vmode, vmode
```

Comment:
We return both some `expected_mode` and some `Value.lr`, backed by the the same variable.

We want to encode extra position information in the former. 
We want the latter to the both left and right because it will be used so.